src/devices_models/devices/common/device_range_constraints.jl

```diff
@@ -0,0 +1,172 @@
+    constraint_name,
+    variable_name,
+    bin_variable_name = nothing,
+    limits_func,
+    constraint_func,
+)
+    return RangeConstraintInputs(
+        constraint_name,
+        variable_name,
+        bin_variable_name,
+        limits_func,
+        constraint_func,
+    )
+end
+
+struct TimeSeriesConstraintInputs
+    constraint_name::String
+    variable_name::String
+    bin_variable_name::Union{Nothing, String}
+    parameter_name::Union{Nothing, String}
+    forecast_label::Union{Nothing, String}
+    multiplier_func::Union{Nothing, Function}
+    constraint_func::Function
+end
+
+function TimeSeriesConstraintInputs(;
+    constraint_name,
+    variable_name,
+    bin_variable_name = nothing,
+    parameter_name,
+    forecast_label,
+    multiplier_func,
+    constraint_func,
+)
+    return TimeSeriesConstraintInputs(
+        constraint_name,
+        variable_name,
+        bin_variable_name,
+        parameter_name,
+        forecast_label,
+        multiplier_func,
+        constraint_func,
+    )
+end
+
+struct DeviceRangeConstraintInputs
+    range_constraint_inputs::Vector{RangeConstraintInputs}
+    timeseries_range_constraint_inputs::Vector{TimeSeriesConstraintInputs}
+    custom_psi_container_func::Union{Nothing, Function}
+    devices_filter_func::Union{Nothing, Function}
+end
+
+"""
+Construct inputs for creating range constraints.
+
+# Arguments
+`range_constraint_inputs::Vector{RangeConstraintInputs}`: May be emtpy.
+`timeseries_range_constraint_inputs::Vector{TimeSeriesConstraintInputs}`: May be empty.
+`custom_psi_container_func::Union{Nothing, Function}`: Optional function to add custom
+ constraints to the internals of a PSIContainer. Must accept PSIContainer, devices iterable,
+ and a subtype of AbstractDeviceFormulation.
+`devices_filter_func::Union{Nothing, Function}`: Optional function to filter the devices on
+
+"""
+function DeviceRangeConstraintInputs(;
+    range_constraint_inputs = Vector{RangeConstraintInputs}(),
+    timeseries_range_constraint_inputs = Vector{TimeSeriesConstraintInputs}(),
+    custom_psi_container_func = nothing,
+    devices_filter_func = nothing,
+)
+    return DeviceRangeConstraintInputs(
+        range_constraint_inputs,
+        timeseries_range_constraint_inputs,
+        custom_psi_container_func,
+        devices_filter_func,
+    )
+end
+
+function device_range_constraints!(
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{T},
+    model::DeviceModel{T, U},
+    feedforward::Union{Nothing, AbstractAffectFeedForward},
+    inputs::DeviceRangeConstraintInputs,
+) where {T <: PSY.Device, U <: AbstractDeviceFormulation}
+    range_constraints = inputs.range_constraint_inputs
+    timeseries_range_constraints = inputs.timeseries_range_constraint_inputs
+    custom_psi_container_func = inputs.custom_psi_container_func
+
+    if !isnothing(inputs.devices_filter_func)
+        devices = filter!(inputs.devices_filter_func, collect(devices))
```

Comment:
It seems we are incurring an extra cost by collecting here the devices. It makes the use of the iterator innecessary.